### ISSUE TITLE
Tabs 组件受控模式下, 设置 activeKey 时, 下一个事件循环再更新 state 时导致的 InkTabBarNode 多次设置样式 BUG

### ISSUE THREAD

**Issue by raohong** [ant-design/ant-design#23386]
- [ ] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/nostalgic-bassi-o775i)

### Steps to reproduce
可以看到切换时下划线不自然. 发现是没有做 activeKey 比较导致重复设置
或者查看 https://preview.pro.ant.design/list/search/projects
更详细重现:  https://github.com/raohong/rc-tabs-bug.git

### What is expected?
修复

### What is actually happening?
切换时下划线不自然, 多次设置

| Environment | Info |
|---|---|
| antd | 4.1.2 |
| React | 16.8 |
| System | windows |
| Browser | chrome |

**Comment by lovelyJason** [comment]
楼主,我们也是这个问题,遍寻无果,没有好的解决方式.官方大大们给个解释呀

**Comment by yoyo837** [comment]
是有那么小顿挫，长短不同的时候比较容易看出来。

**Comment by lovelyJason** [comment]
> 是有那么小顿挫，长短不同的时候比较容易看出来。

不是吧,这个跟render次数有关的.比如发三次网络请求setState,下划线原地左右抽搐三次,再移动到最终的tab,和宽度没啥关系

**Comment by raohong** [comment]
> 楼主,我们也是这个问题,遍寻无果,没有好的解决方式.官方大大们给个解释呀

目前我自己改了一下它的源码, 
```
function _componentDidUpdate(component, init, prev) {
  var _component$props = component.props,
      styles = _component$props.styles,
      panels = _component$props.panels,
      activeKey = _component$props.activeKey,
      direction = _component$props.direction;
  var rootNode = component.props.getRef('root');
  var wrapNode = component.props.getRef('nav') || rootNode;
  var inkBarNode = component.props.getRef('inkBar');
  var activeTab = component.props.getRef('activeTab');
  var inkBarNodeStyle = inkBarNode.style;
  var tabBarPosition = component.props.tabBarPosition;
  var activeIndex = getActiveIndex(panels, activeKey);

  if (init) {
    // prevent mount animation
    inkBarNodeStyle.display = 'none';
  }

  if (activeTab && (!prev ||  activeKey !== prev.activeKey)) {
    var tabNode = activeTab;
    var transformSupported = isTransform3dSupported(inkBarNodeStyle); // Reset current style

    setTransform(inkBarNodeStyle, '');
    inkBarNodeStyle.width = '';
    inkBarNodeStyle.height = '';
    inkBarNodeStyle.left = '';
    inkBarNodeStyle.top = '';
    inkBarNodeStyle.bottom = '';
    inkBarNodeStyle.right = '';


    if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
      var left = getLeft(tabNode, wrapNode);
      var width = tabNode.offsetWidth; // If tabNode'width width equal to wrapNode'width when tabBarPosition is top or bottom
      // It means no css working, then ink bar should not have width until css is loaded
      // Fix https://github.com/ant-design/ant-design/issues/7564

      if (width === rootNode.offsetWidth) {
        width = 0;
      } else if (styles.inkBar && styles.inkBar.width !== undefined) {
        width = parseFloat(styles.inkBar.width, 10);

        if (width) {
          left += (tabNode.offsetWidth - width) / 2;
        }
      }

      if (direction === 'rtl') {
        left = getStyle(tabNode, 'margin-left') - left;
      } // use 3d gpu to optimize render


      if (transformSupported) {
        setTransform(inkBarNodeStyle, "translate3d(".concat(left, "px,0,0)"));
      } else {
        inkBarNodeStyle.left = "".concat(left, "px");
      }

      inkBarNodeStyle.width = "".concat(width, "px");
    } else {
      var top = getTop(tabNode, wrapNode, true);
      var height = tabNode.offsetHeight;

      if (styles.inkBar && styles.inkBar.height !== undefined) {
        height = parseFloat(styles.inkBar.height, 10);

        if (height) {
          top += (tabNode.offsetHeight - height) / 2;
        }
      }

      if (transformSupported) {
        setTransform(inkBarNodeStyle, "translate3d(0,".concat(top, "px,0)"));
        inkBarNodeStyle.top = '0';
      } else {
        inkBarNodeStyle.top = "".concat(top, "px");
      }

      inkBarNodeStyle.height = "".concat(height, "px");
    }
  }

  inkBarNodeStyle.display = activeIndex !== -1 ? 'block' : 'none';
}
```

```
{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prev) {
      _componentDidUpdate(this, false, prev);
    }
  }
```

**Comment by lovelyJason** [comment]
> > 楼主,我们也是这个问题,遍寻无果,没有好的解决方式.官方大大们给个解释呀
> 
> 目前我自己改了一下它的源码,
> 
> ```
> function _componentDidUpdate(component, init, prev) {
>   var _component$props = component.props,
>       styles = _component$props.styles,
>       panels = _component$props.panels,
>       activeKey = _component$props.activeKey,
>       direction = _component$props.direction;
>   var rootNode = component.props.getRef('root');
>   var wrapNode = component.props.getRef('nav') || rootNode;
>   var inkBarNode = component.props.getRef('inkBar');
>   var activeTab = component.props.getRef('activeTab');
>   var inkBarNodeStyle = inkBarNode.style;
>   var tabBarPosition = component.props.tabBarPosition;
>   var activeIndex = getActiveIndex(panels, activeKey);
> 
>   if (init) {
>     // prevent mount animation
>     inkBarNodeStyle.display = 'none';
>   }
> 
>   if (activeTab && (!prev ||  activeKey !== prev.activeKey)) {
>     var tabNode = activeTab;
>     var transformSupported = isTransform3dSupported(inkBarNodeStyle); // Reset current style
> 
>     setTransform(inkBarNodeStyle, '');
>     inkBarNodeStyle.width = '';
>     inkBarNodeStyle.height = '';
>     inkBarNodeStyle.left = '';
>     inkBarNodeStyle.top = '';
>     inkBarNodeStyle.bottom = '';
>     inkBarNodeStyle.right = '';
> 
> 
>     if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
>       var left = getLeft(tabNode, wrapNode);
>       var width = tabNode.offsetWidth; // If tabNode'width width equal to wrapNode'width when tabBarPosition is top or bottom
>       // It means no css working, then ink bar should not have width until css is loaded
>       // Fix https://github.com/ant-design/ant-design/issues/7564
> 
>       if (width === rootNode.offsetWidth) {
>         width = 0;
>       } else if (styles.inkBar && styles.inkBar.width !== undefined) {
>         width = parseFloat(styles.inkBar.width, 10);
> 
>         if (width) {
>           left += (tabNode.offsetWidth - width) / 2;
>         }
>       }
> 
>       if (direction === 'rtl') {
>         left = getStyle(tabNode, 'margin-left') - left;
>       } // use 3d gpu to optimize render
> 
> 
>       if (transformSupported) {
>         setTransform(inkBarNodeStyle, "translate3d(".concat(left, "px,0,0)"));
>       } else {
>         inkBarNodeStyle.left = "".concat(left, "px");
>       }
> 
>       inkBarNodeStyle.width = "".concat(width, "px");
>     } else {
>       var top = getTop(tabNode, wrapNode, true);
>       var height = tabNode.offsetHeight;
> 
>       if (styles.inkBar && styles.inkBar.height !== undefined) {
>         height = parseFloat(styles.inkBar.height, 10);
> 
>         if (height) {
>           top += (tabNode.offsetHeight - height) / 2;
>         }
>       }
> 
>       if (transformSupported) {
>         setTransform(inkBarNodeStyle, "translate3d(0,".concat(top, "px,0)"));
>         inkBarNodeStyle.top = '0';
>       } else {
>         inkBarNodeStyle.top = "".concat(top, "px");
>       }
> 
>       inkBarNodeStyle.height = "".concat(height, "px");
>     }
>   }
> 
>   inkBarNodeStyle.display = activeIndex !== -1 ? 'block' : 'none';
> }
> ```
> 
> ```
> {
>     key: "componentDidUpdate",
>     value: function componentDidUpdate(prev) {
>       _componentDidUpdate(this, false, prev);
>     }
>   }
> ```

我的node_modules里面的tabs代码怎么跟你这个长的不一样

**Comment by raohong** [comment]
> > > 楼主,我们也是这个问题,遍寻无果,没有好的解决方式.官方大大们给个解释呀
> > 
> > 
> > 目前我自己改了一下它的源码,
> > ```
> > function _componentDidUpdate(component, init, prev) {
> >   var _component$props = component.props,
> >       styles = _component$props.styles,
> >       panels = _component$props.panels,
> >       activeKey = _component$props.activeKey,
> >       direction = _component$props.direction;
> >   var rootNode = component.props.getRef('root');
> >   var wrapNode = component.props.getRef('nav') || rootNode;
> >   var inkBarNode = component.props.getRef('inkBar');
> >   var activeTab = component.props.getRef('activeTab');
> >   var inkBarNodeStyle = inkBarNode.style;
> >   var tabBarPosition = component.props.tabBarPosition;
> >   var activeIndex = getActiveIndex(panels, activeKey);
> > 
> >   if (init) {
> >     // prevent mount animation
> >     inkBarNodeStyle.display = 'none';
> >   }
> > 
> >   if (activeTab && (!prev ||  activeKey !== prev.activeKey)) {
> >     var tabNode = activeTab;
> >     var transformSupported = isTransform3dSupported(inkBarNodeStyle); // Reset current style
> > 
> >     setTransform(inkBarNodeStyle, '');
> >     inkBarNodeStyle.width = '';
> >     inkBarNodeStyle.height = '';
> >     inkBarNodeStyle.left = '';
> >     inkBarNodeStyle.top = '';
> >     inkBarNodeStyle.bottom = '';
> >     inkBarNodeStyle.right = '';
> > 
> > 
> >     if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
> >       var left = getLeft(tabNode, wrapNode);
> >       var width = tabNode.offsetWidth; // If tabNode'width width equal to wrapNode'width when tabBarPosition is top or bottom
> >       // It means no css working, then ink bar should not have width until css is loaded
> >       // Fix https://github.com/ant-design/ant-design/issues/7564
> > 
> >       if (width === rootNode.offsetWidth) {
> >         width = 0;
> >       } else if (styles.inkBar && styles.inkBar.width !== undefined) {
> >         width = parseFloat(styles.inkBar.width, 10);
> > 
> >         if (width) {
> >           left += (tabNode.offsetWidth - width) / 2;
> >         }
> >       }
> > 
> >       if (direction === 'rtl') {
> >         left = getStyle(tabNode, 'margin-left') - left;
> >       } // use 3d gpu to optimize render
> > 
> > 
> >       if (transformSupported) {
> >         setTransform(inkBarNodeStyle, "translate3d(".concat(left, "px,0,0)"));
> >       } else {
> >         inkBarNodeStyle.left = "".concat(left, "px");
> >       }
> > 
> >       inkBarNodeStyle.width = "".concat(width, "px");
> >     } else {
> >       var top = getTop(tabNode, wrapNode, true);
> >       var height = tabNode.offsetHeight;
> > 
> >       if (styles.inkBar && styles.inkBar.height !== undefined) {
> >         height = parseFloat(styles.inkBar.height, 10);
> > 
> >         if (height) {
> >           top += (tabNode.offsetHeight - height) / 2;
> >         }
> >       }
> > 
> >       if (transformSupported) {
> >         setTransform(inkBarNodeStyle, "translate3d(0,".concat(top, "px,0)"));
> >         inkBarNodeStyle.top = '0';
> >       } else {
> >         inkBarNodeStyle.top = "".concat(top, "px");
> >       }
> > 
> >       inkBarNodeStyle.height = "".concat(height, "px");
> >     }
> >   }
> > 
> >   inkBarNodeStyle.display = activeIndex !== -1 ? 'block' : 'none';
> > }
> > ```
> > 
> > 
> > ```
> > {
> >     key: "componentDidUpdate",
> >     value: function componentDidUpdate(prev) {
> >       _componentDidUpdate(this, false, prev);
> >     }
> >   }
> > ```
> 
> 我的node_modules里面的tabs代码怎么跟你这个长的不一样

是在 rc-tabs 里 的 es

**Comment by yoyo837** [comment]
> > 是有那么小顿挫，长短不同的时候比较容易看出来。
> 
> 不是吧,这个跟render次数有关的.比如发三次网络请求setState,下划线原地左右抽搐三次,再移动到最终的tab,和宽度没啥关系

你的问题整一个复现看看，楼主的复现看上去不是这样。

**Comment by lovelyJason** [comment]
果然是这里源码的问题,我也改了源码了

**Comment by yoyo837** [comment]
来个PR？

**Comment by LbhFront-end** [comment]
同样的问题 请问解决了吗 换了最新版本4.25 看更新记录 好像没有还没有

**Comment by yoyo837** [comment]
通常已解决的问题会体现在issue关闭状态或者关联的PR上。

**Comment by afc163** [comment]
It should have been fixed in 4.3.x now: https://codesandbox.io/s/lively-water-imvyq?file=/src/App.js
